Flow.Launcher/ViewModel/MainViewModel.cs

```diff
@@ -189,6 +193,51 @@ private void InitializeKeyCommands()
 
             SelectFirstResultCommand = new RelayCommand(_ => SelectedResults.SelectFirstResult());
 
+            ReverseHistory = new RelayCommand(_ => {
+
+                if (_history.Items.Count > 0)
+                {
+                    ChangeQueryText(_history.Items[_history.Items.Count - _lastHistoryIndex].Query.ToString());
+
+                    if (_lastHistoryIndex < _history.Items.Count)
+                    {
+                        _lastHistoryIndex++;
+                    }
+                }
+            });
+
+            ForwardHistory = new RelayCommand(_ => {
+
+                if (_history.Items.Count > 0)
+                {
+                    ChangeQueryText(_history.Items[_history.Items.Count - _lastHistoryIndex].Query.ToString());
+
+                    if (_lastHistoryIndex > 1)
+                    {
+                        _lastHistoryIndex--;
+                    }
+                }
+                
+            });
+
+            InsertLastQuery = new RelayCommand(_ => {
+                var results = SelectedResults;
+
+                if (_history.Items.Count > 0 
+                    && _queryText == String.Empty
+                    && !HistorySelected() 
+                    && !ContextMenuSelected())
+                {
+                    ChangeQueryText(_history.Items[_history.Items.Count - 1].Query.ToString());
+                    _lastHistoryIndex = 2;
```

Comment:
why 2? this is the cause for the above error in the comment. try not to use plain numbers and strings please

Reply:
Oh this is set so if the user hits <kbd>Up</kbd> on an empty Query and decides to use <kbd>Alt</kbd>+<kbd>Up</kbd> they don't repeat inserting the first item.

Reply:
I should add logic to make sure we dont set this to an index that doesn't exist

Reply:
Fixed: https://github.com/Flow-Launcher/Flow.Launcher/pull/852/commits/c6d5525dec46631db6cf587573d6ecc55c8b9419